snuba/clickhouse/http.py

```diff
@@ -73,15 +94,80 @@ def write(self, rows: Iterable[WriterTableRow]):
             chunked=True,
         )
 
-        if response.status != 200:
-            # XXX: This should be switched to just parse the JSON body after
-            # https://github.com/yandex/ClickHouse/issues/6272 is available.
-            content = response.data.decode("utf8")
-            details = CLICKHOUSE_ERROR_RE.match(content)
-            if details is not None:
-                code, type, message = details.groups()
-                raise ClickhouseError(int(code), message)
-            else:
-                raise HTTPError(
-                    f"Received unexpected {response.status} response: {content}"
-                )
+        raise_for_error_response(response)
+
+
+def parse_and_transform_date(value: str) -> str:
+    return transform_date(dateutil_parse(value))
+
+
+def parse_and_transform_datetime(value: str) -> str:
+    return transform_datetime(dateutil_parse(value))
+
+
+transform_column_types = build_result_transformer(
+    [
+        (re.compile(r"^Date(\(.+\))?$"), parse_and_transform_date),
+        (re.compile(r"^DateTime(\(.+\))?$"), parse_and_transform_datetime),
+    ]
+)
+
+
+class HTTPReader(Reader[ClickhouseQuery]):
+    def __init__(
+        self, host: str, port: int, settings: Optional[Mapping[str, str]] = None
+    ):
+        if settings is not None:
+            assert "query_id" not in settings, "query_id cannot be passed as a setting"
+
+        self.__pool = HTTPConnectionPool(host, port)
+
+        self.__default_settings: MutableMapping[str, str] = (
+            {**settings} if settings is not None else {}
+        )
+
+        if "output_format_json_quote_64bit_integers" not in self.__default_settings:
+            self.__default_settings["output_format_json_quote_64bit_integers"] = "0"
+
+    def execute(
+        self,
+        query: ClickhouseQuery,
+        settings: Optional[Mapping[str, str]] = None,
+        query_id: Optional[str] = None,
+        with_totals: bool = False,  # NOTE: unnecessary with FORMAT JSON
+    ) -> Result:
+        query_settings: MutableMapping[str, str] = (
+            {**settings} if settings is not None else {}
+        )
+
+        # XXX: mypy won't allow redefining ``settings`` as mutable, so delete
+        # the original variable to avoid accidentally referencing ``settings``
+        # instead of ``query_settings``.
+        del settings
+
+        assert (
+            "query_id" not in query_settings
+        ), "query_id cannot be passed as a setting"
+
+        if query_id is not None:
+            query_settings["query_id"] = query_id
+
+        response = self.__pool.urlopen(
+            "POST",
+            "/?" + urlencode({**self.__default_settings, **query_settings}),
+            headers={"Connection": "keep-alive", "Accept-Encoding": "gzip,deflate"},
+            body=query.format_sql("JSON"),
+        )
```

Comment:
I think it may be useful to set up retries for network connection errors. Also what's the default timeout we would have here?

Reply:
> I think it may be useful to set up retries for network connection errors.

Makes sense, will look into that.

> Also what's the default timeout we would have here?

Not sure. Will investigate tomorrow.

Reply:
Context on how we handled the retries and timeout in the queries from sentry to snuba with the reasoning around retries if it helps 
https://github.com/getsentry/sentry/commit/6d5486696dbd958fe20c3a6180b925468455257d

Reply:
> I think it may be useful to set up retries for network connection errors.

I started writing a comment about how on second thought, I didn't think we should just blindly retry on errors without a reason to, but then I checked the distinction between `execute` and `execute_robust` in [`ClickhousePool`](https://github.com/getsentry/snuba/blob/07cd406d5cda20bf78485627abd2743ae0fa2e05/snuba/clickhouse/native.py) and I guess we retry there as well.

> Also what's the default timeout we would have here?

This basically falls through to to `socket.getdefaulttimeout`:

```
(snuba) ted@veneno % docker run --rm -it getsentry/snuba python -c 'import socket; print(socket.getdefaulttimeout())'
+ '[' python = bash ']'
+ '[' p = - ']'
+ '[' python = api ']'
+ snuba python --help
+ exec gosu snuba python -c 'import socket; print(socket.getdefaulttimeout())'
None
```

I guess this should be able to be parameterized (the writer also has this issue.)

Reply:
About the retries, the difference between execute_robust and using the default Retry object of urllib3 is that execute_robust also waits by default 1 second (besides being more flexible on the type of errors).
I would say we should always retry (a given amount of times) for errors when establishing the connection (NetworkError, SocketTimeout, etc.) That covers the transient network issues that we run into. In these case you will get the Exception from urllib.
Whether to retry on ServerError that is debatable. Here (this is only a read path so idempotent) it is not dangerous but it puts useless load to the DB if we do that on errors that cannot be transient. I am not sure we have a clean way to make this distinction on clickhouse ServerErrors, unless the HTTP reader clickhosue used HTTP 5xx for actual server errors and HTTP 4xx for query errors. If it does not we would have to selectively blacklist/whitelist error codes I guess.
I'd be ok being conservative and not retry at all for these cases. (Sentry retries them by the way).
I would certainly not retry on ReadTimeout errors.
How does it sound as a strategy?

Reply:
Actually the Retry object from urllib3 is already capable of making the distinction between network errors, read errors and redirect errors so you do not need to implement the logic in your code to do network connection retries:
https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry

Reply:
> About the retries, the difference between execute_robust and using the default Retry object of urllib3 is that execute_robust also waits by default 1 second (besides being more flexible on the type of errors).

Yeah, my original comment was unclear — I meant that i was surprised that `execute` also had some number of baked in retries. (That doesn't really change anything about your comment, though.) It makes some more sense why after this conversation.

> Whether to retry on ServerError that is debatable.

In my opinion, we should try and avoid implicit retries of application operations as much as possible, and allow (well, not so much allow as require) the caller to retry when necessary. I don't think a one-size-fits-all solution in the reader (or writer, or connection pool, etc.) is a good idea.

Like you mentioned, I do think it makes sense to always retry on transport level errors (connection timeouts.) I don't think of this so much as the responsibility of the `Reader` as the transport itself (specifically the connection pool, native or HTTP) so it probably makes sense to just pass the instantiated pool to the constructor, rather than constructing it in the reader itself.

This also is similar to some of the conversations around #844 around connection sharing. It'd be unnecessarily wasteful to create a new connection pool for every reader if they are all communicating with the same cluster.

Reply:
I would be totally ok if we decided to retry only on transport errors. I am not a big fan of baking application specific retries in the Reader (or lower) layer either.